lumicks/pylake/confocal.py

```diff
@@ -0,0 +1,232 @@
+import json
+import numpy as np
+
+from .detail.mixin import PhotonCounts
+from .detail.mixin import ExcitationLaserPower
+from .detail.image import save_tiff, ImageMetadata
+
+
+class BaseScan(PhotonCounts, ExcitationLaserPower):
+    """Base class for confocal scans
+
+    Parameters
+    ----------
+    name : str
+        confocal scan name
+    file : lumicks.pylake.File
+        Parent file. Contains the channel data.
+    start : int
+        Start point in the relevant info wave.
+    stop : int
+        End point in the relevant info wave.
+    json : dict
+        Dictionary containing metadata.
+    """
+
+    def __init__(self, name, file, start, stop, json):
+        self.start = start
+        self.stop = stop
+        self.name = name
+        self.json = json
+        self.file = file
+        self._cache = {}
+
+    @classmethod
+    def from_dataset(cls, h5py_dset, file):
+        """
+        Construct a confocal class from dataset.
+
+        Parameters
+        ----------
+        h5py_dset : h5py.Dataset
+            The original HDF5 dataset containing confocal scan information
+        file : lumicks.pylake.File
+            The parent file. Used to loop up channel data
+        """
+        start = h5py_dset.attrs["Start time (ns)"]
+        stop = h5py_dset.attrs["Stop time (ns)"]
+        name = h5py_dset.name.split("/")[-1]
+        try:
+            json_data = json.loads(h5py_dset[()])["value0"]
+        except KeyError:
+            # TODO => use class name
+            raise KeyError(f"Scan '{name}' is missing metadata and cannot be loaded")
+
+        return cls(name, file, start, stop, json_data)
+
+    def _get_photon_count(self, name):
+        # Grab the portion of the photon count that overlaps with the scan
+        photon_count = getattr(self.file, f"{name}_photon_count".lower())[self.start:self.stop]
+        timeline_start = photon_count._src.start
+        timeline_dt = photon_count._src.dt
+
+        # Workaround for a bug in the STED delay mechanism which could result in scan start times ending up within
+        # the sample time.
+        if timeline_start - timeline_dt < self.start < timeline_start:
+            self.start = timeline_start
+
+        # Checks whether the scan or kymograph starts before the timeline information. 
+        # If this is the case, it will lead to an incorrect reconstruction. 
+        # If implemented, resolve the problem.
+        if timeline_start > self.start:
+            self._fix_incorrect_start()
+
+        return getattr(self.file, f"{name}_photon_count".lower())[self.start:self.stop]
```

Comment:
This ends up reading the photon counts twice, even if there's no need to fix the incorrect start.